### PR TITLE
[GLib] D-Bus proxy quietly fails if host session bus address is an abstract socket

### PR DESCRIPTION
**Pull request by mcatanzaro**
#### 79f41316d99e6496904950be78a0015503385ec2
<pre>
[GLib] D-Bus proxy quietly fails if host session bus address is an abstract socket
<a href="https://bugs.webkit.org/show_bug.cgi?id=245843">https://bugs.webkit.org/show_bug.cgi?id=245843</a>

Reviewed by Carlos Garcia Campos.

Nowadays all major Linux distros run the D-Bus session bus using a
standard Unix socket created on the filesystem, but distros that do not
use systemd still wind up using dbus-daemon&apos;s default session bus
address, which up until now has used the abstract socket namespace.

Our code here is only compatible with filesystem sockets since it
attempts to create the proxy bus socket in the sandbox at exactly the
same location within the sandbox that the real session bus socket exists
on the host system. If the host session bus uses an abstract socket, our
code just fails. There&apos;s no particular reason to do things this way, so
let&apos;s not. Instead, we&apos;ll always create the proxy bus socket in a
well-known location within the sandbox, /run/webkitgtk/bus or
/run/wpe/bus. This matches flatpak&apos;s behavior and should allow things to
work regardless.

The accessibility bus requires the same changes.

Note there are major security problems if the host session bus uses an
abstract socket. See <a href="https://gitlab.freedesktop.org/dbus/dbus/-/issues/416">https://gitlab.freedesktop.org/dbus/dbus/-/issues/416</a>
for full details. While this configuration is not recommended, it&apos;s
usually safe for WebKit because our sandbox does not allow network access
(unless using a non-local X server, which is inherently insecure anyway).

* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::bindDBusSession):
(WebKit::bindA11y):
* Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp:
(WebKit::XDGDBusProxy::dbusSessionPath):
(WebKit::XDGDBusProxy::accessibilityPath):
(WebKit::XDGDBusProxy::dbusSessionProxy):
(WebKit::XDGDBusProxy::accessibilityProxy):
(WebKit::XDGDBusProxy::makePath): Deleted.
* Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.h:

Canonical link: <a href="https://commits.webkit.org/255218@main">https://commits.webkit.org/255218@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c26298eec7436f5d085afa3a364db526f6d428fe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91636 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/865 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22239 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101334 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/863 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29472 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83953 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97669 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97294 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/498 "Found 1 new test failure: fast/forms/ios/file-upload-panel-capture.html (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78295 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27457 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82422 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70494 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35758 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16097 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33513 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17193 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3618 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37354 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39952 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39259 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36310 "Passed tests") | |

### CODE REVIEW

**Comment by nater1983** [comment]
Which version of Webkit does this patch against, i tried against 2.38.0 and it failed

File: Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp]: This is leaking, g_build_filename returns a new allocated string and CString copies the given string. Why does this belong here? it's not using anything from the XDGDbusProxy, it's the path where the launcher wants to bind the socket, so it can be moved there using GUniquePtr<char> instead of CString. I'm not sure it's worth caching it, since the dbus proxy is only launched once.

**Review by carlosgcampos — CHANGES_REQUESTED** [review]

**Comment by mcatanzaro** [comment]
> Which version of Webkit does this patch against, i tried against 2.38.0 and it failed

It's a pull request against the main branch. But you _might_ be able to apply it on 2.38.0 if you take f6198cdad380b68e02c7a135b18fcf7e387078ea first.

File: Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp
Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp]: > This is leaking, g_build_filename returns a new allocated string and CString copies the given string.

Oops. :S

> Why does this belong here? it's not using anything from the XDGDbusProxy, it's the path where the launcher wants to bind the socket, so it can be moved there using GUniquePtr instead of CString. 

The session bus path can be moved to BubblewrapLauncher.cpp, indeed. I like that idea. Except: it's logically paired with the accessibility bus path, and moving that would be more awkward due to the call to PlatformDisplay::setAccessibilityBusAddress. That call really wants to stay in XDGDBusProxy.cpp because we want to call it only once, when the accessibility proxy is actually launched. I suppose it would be harmless to call it repeatedly, but that's not quite as nice. I suppose we could have XDGDBusProxy::accessibilityProxy accept the path as a parameter, so that it can be passed to PlatformDisplay::setAccessibilityBusAddress that way. Sound good?

> I'm not sure it's worth caching it, since the dbus proxy is only launched once.

But the path is needed once per process launched, it's not actually needed when the proxy itself is launched. And the path never changes. Caching is not important though, just a very tiny optimization.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp]: > I suppose we could have XDGDBusProxy::accessibilityProxy accept the path as a parameter, so that it can be passed to PlatformDisplay::setAccessibilityBusAddress that way. Sound good?

This actually works pretty well.

**Comment by nater1983** [comment]
it works here

**Comment by nater1983** [comment]
image

**Comment by oreo639** [comment]
I can confirm that this PR works for me as well.
Although I did run into an unrelated build failure due to `${CMAKE_DL_LIBS}` missing here: https://github.com/WebKit/WebKit/blob/d257ea20c5996ada3c009a6a2f4a639f92c0e2ca/Source/ThirdParty/ANGLE/CMakeLists.txt#L145

(It doesn't affect glibc versions after 2.34 due to libdl being integrated into the libc)

**Review by carlosgcampos — APPROVED** [review]

**Comment by nekopsykose** [comment]
this seems to work only before the dbus-1.15.2 change to make non-abstract sockets the default (and only option), where it fails again in the same way as before

**Comment by mcatanzaro** [comment]
> this seems to work only before the dbus-1.15.2 change to make non-abstract sockets the default (and only option), where it fails again in the same way as before

Really? But non-abstract sockets are default in all major distros and work fine currently. So we're going to need more info on what's going wrong. I guess start with a bug report....

**Comment by mcatanzaro** [comment]
(BTW the D-Bus 1.15.2 should be enough to fix this bug on its own without any changes in WebKit.)

**Comment by nekopsykose** [comment]
yeah, i know it's the default in most places.. i just mean in this same context that failed to begin with, on a distro without systemd (alpine linux), in a session started with dbus-run-session. to my knowledge, this always worked before too (these fixes didn't affect systemd distros with "modern" dbus sessions, for instance).  

i can open a separate bug report if you'd like, but it's pretty much the same thing as the epiphany thread. (just with dbus-run-session from dbus 1.15.2, and 2.38.0 webkit with this mr and the linked commit above backported. it worked with abstract-dbus in 1.15.0)

**Comment by mcatanzaro** [comment]
> i can open a separate bug report if you'd like, but it's pretty much the same thing as the epiphany thread.

I definitely need a new bug report, because I don't know what could possibly be wrong this time.

**Comment by nekopsykose** [comment]
sure, where should i open it?  

@oreo639 could you update to the latest dbus from today with the socket changes and test again? i have a hunch it won't work again, but if it does it would be some other issue

**Comment by mcatanzaro** [comment]
> sure, where should i open it?

WebKit Bugzilla, WebKitGTK component. Please include the output of `echo $DBUS_SESSION_BUS_ADDRESS` so I can see specifically what it's trying to connect to. Thanks.

P.S. Noticing a regression only 5 minutes after the patch landed was pretty good response time... that deserves some sort of award. I had been planning to backport this to webkit-2.38, but now I will hold off until we understand what is wrong, since the D-Bus release is already out and that should be enough to fix the original problem for affected users.

**Comment by nekopsykose** [comment]
> P.S. Noticing a regression only 5 minutes after the patch landed was pretty good response time... that deserves some sort of award. I had been planning to backport this to webkit-2.38, but now I will hold off until we understand what is wrong, since the D-Bus release is already out and that should be enough to fix the original problem for affected users.  

i had started testing it before it was merged :) but to be clear, this isn't a regression. epiphany exhibits the same crash loop before these changes- they were originally needed to fix it with the abstract socket, and now with only the non-abstract one being available it fails anyway. i will put all the info in the report, give me a moment

**Comment by nekopsykose** [comment]
(and to be clear- if i can only reproduce this with an epiphany startup, the report still goes there, right? i don't know how i'd test it otherwise)

**Comment by mcatanzaro** [comment]
Yes, use WebKit Bugzilla please.

**Comment by nekopsykose** [comment]
https://bugs.webkit.org/show_bug.cgi?id=246159 done

**Comment by oreo639** [comment]
> could you update to the latest dbus from today with the socket changes and test again? i have a hunch it won't work again, but if it does it would be some other issue

Arg, yeah, good point. xdg-dbus-proxy can't see the dbus session address if it is a file in `/tmp`, since it just mounts `/run` assuming it is somewhere in there.

**Comment by nekopsykose** [comment]
> since it just mounts /run assuming it is somewhere in there.  

which works on every "modern" dbus since they seem to put it under the xdg_runtime_dir present in /run/user/$uid . i do wonder though why dbus-run-session/dbus-launch doesn't do the same, when the runtime dir environment variable is present where it's invoked from..

**Comment by mcatanzaro** [comment]
> Arg, yeah, good point. xdg-dbus-proxy can't see the dbus session address if it is a file in /tmp, since it just mounts /run assuming it is somewhere in there.

Oh, oops, OK... good catch. That's definitely the problem.

**Comment by oreo639** [comment]
The following patch works for me:
```
diff --git a/Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp b/Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp
index 376862827cd0..1a9f0ae9d54b 100644
--- a/Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp
+++ b/Source/WebKit/UIProcess/Launcher/glib/XDGDBusProxy.cpp
@@ -59,6 +59,23 @@ CString XDGDBusProxy::makeProxy(const char* proxyTemplate)
     return proxySocketTemplate.get();
 }
 
+CString XDGDBusProxy::makePath(const char* dbusAddress)
+{
+    if (!dbusAddress || !g_str_has_prefix(dbusAddress, "unix:"))
+        return { };
+
+    if (const char* path = strstr(dbusAddress, "path=")) {
+        path += strlen("path=");
+        const char* pathEnd = path;
+        while (*pathEnd && *pathEnd != ',')
+            pathEnd++;
+
+        return CString(path, pathEnd - path);
+    }
+
+    return { };
+}
+
 std::optional<CString> XDGDBusProxy::dbusSessionProxy(AllowPortals allowPortals)
 {
     if (!m_dbusSessionProxyPath.isNull())
@@ -184,6 +201,12 @@ bool XDGDBusProxy::launch()
     ProcessLauncher::LaunchOptions launchOptions;
     launchOptions.processType = ProcessLauncher::ProcessType::DBusProxy;
 
+    const char* dbusAddress = g_getenv("DBUS_SESSION_BUS_ADDRESS");
+    CString dbusAddressPath = makePath(dbusAddress);
+
+    if (!dbusAddressPath.isNull())
+        launchOptions.extraSandboxPaths.add(dbusAddressPath, SandboxPermission::ReadWrite);
+
     GUniqueOutPtr<GError> error;
     GRefPtr<GSubprocess> process = bubblewrapSpawn(launcher.get(), launchOptions, argv, &error.outPtr());
     if (!process)
```

**Comment by nater1983** [comment]
I confirm oreo's patch work's on my host system i am no longer using Malloc=1.
website https://www.heise.de will not load still. Not sure if there is other websites. 
```
bash-5.2# export
declare -x COLORTERM="truecolor"
declare -x DBUS_SESSION_BUS_ADDRESS="unix:path=/tmp/dbus-oJ4v9WbAfI,guid=dbaf0c654d0055840fdff197633fa4e6"
declare -x DESKTOP_SESSION="gnome-xorg"
declare -x DISPLAY=":0"
declare -x GDK_USE_XFT="1"
declare -x GDMSESSION="gnome-xorg"
declare -x GDM_LANG="C"
declare -x GNOME_KEYRING_CONTROL="/home/test/.cache/keyring-ISQHT1"
declare -x GNOME_TERMINAL_SCREEN="/org/gnome/Terminal/screen/7cc56ae0_eadc_4032_914b_95af0b87a842"
declare -x GNOME_TERMINAL_SERVICE=":1.66"
declare -x G_BROKEN_FILENAMES="1"
declare -x G_FILENAME_ENCODING="@locale"
declare -x HOME="/root"
declare -x HOSTNAME="darkstar.example.net"
declare -x INPUTRC="/etc/inputrc"
declare -x LANG="en_US.UTF-8"
declare -x LC_COLLATE="C"
declare -x LESSOPEN="|lesspipe.sh %s"
declare -x LOGNAME="root"
declare -x LS_COLORS=""
declare -x LS_OPTIONS="-F -b -T 0 --color=auto"
declare -x MINICOM="-c on"
declare -x NODE_PATH="/usr/lib64/node_modules"
```
unix:path=/tmp/dbus-oJ4v9WbAfI,guid=dbaf0c654d0055840fdff197633fa4e6

**Comment by nater1983** [comment]
image

**Comment by nater1983** [comment]
@mcatanzaro do we apply this on top of patch 5011?

**Comment by mcatanzaro** [comment]
This is 5011. ;)

**Comment by mcatanzaro** [comment]
You can try #5136 to test the fix for D-Bus sockets under /tmp. It's based on oreo639's proposal above.

**Comment by nater1983** [comment]
srry long day

**Pull request merged**
